modyn/evaluator/internal/grpc/evaluator_grpc_servicer.py

```diff
@@ -0,0 +1,223 @@
+            tempdir: temporary directory to store trained models.
+        """
+        super().__init__()
+
+        self._config = config
+        self._base_dir = tempdir
+
+        assert self._base_dir.exists(), f"Temporary Directory {self._base_dir} should have been created."
+
+        self._lock = Lock()
+
+        self._next_evaluation_id = 0
+        self._evaluation_dict: dict[int, EvaluationInfo] = {}
+        self._evaluation_process_dict: dict[int, EvaluationProcessInfo] = {}
+
+        self._storage_address = f"{config['storage']['hostname']}:{config['storage']['port']}"
+        self._model_storage_address = f"{config['model_storage']['hostname']}:{config['model_storage']['port']}"
+        self._metric_manager = MetricManager()
+
+        self._model_storage_stub = EvaluatorGRPCServicer.connect_to_model_storage(self._model_storage_address)
+
+    @staticmethod
+    def connect_to_model_storage(model_storage_address: str) -> ModelStorageStub:
+        model_storage_channel = grpc.insecure_channel(model_storage_address)
+        assert model_storage_channel is not None
+        if not grpc_connection_established(model_storage_channel):
+            raise ConnectionError(
+                f"Could not establish gRPC connection to model storage at address {model_storage_address}."
+            )
+        return ModelStorageStub(model_storage_channel)
+
+    def evaluate_model(self, request: EvaluateModelRequest, context: grpc.ServicerContext) -> EvaluateModelResponse:
+        logger.info("Received evaluate model request.")
+
+        if not hasattr(dynamic_module_import("modyn.models"), request.model_id):
+            logger.error(f"Model {request.model_id} not available!")
+            return EvaluateModelResponse(evaluation_started=False)
+
+        fetch_request = FetchModelRequest(model_id=request.trained_model_id)
+        fetch_resp: FetchModelResponse = self._model_storage_stub.FetchModel(fetch_request)
+
+        if not fetch_resp.success:
+            logger.error(
+                f"Trained model {request.trained_model_id} cannot be fetched from model storage. "
+                f"Evaluation cannot be started."
+            )
+            return EvaluateModelResponse(evaluation_started=False)
+
+        with self._lock:
+            evaluation_id = self._next_evaluation_id
+            self._next_evaluation_id += 1
+
+        local_model_path = self._base_dir / f"trained_model_{evaluation_id}.modyn"
+
+        download_file(
+            hostname=self._config["model_storage"]["hostname"],
+            port=int(self._config["model_storage"]["ftp_port"]),
+            user="modyn",
+            password="modyn",
+            remote_file_path=pathlib.Path(fetch_resp.model_path),
+            local_file_path=local_model_path,
+            callback=get_pretrained_model_callback(logger),
+        )
+
+        logger.info(f"Successfully downloaded trained model {request.trained_model_id} to {local_model_path}.")
```

Comment:
maybe we can move lines 84-110 in another function to make this one shorter?

Reply:
(and then the next lines could be sth like run evaluation etc, so we factor this method here in short method calls)